Add baseline (#975)
zricethezav/gitleaks: README.md

      --redact                 redact secrets from logs and stdout
  -f, --report-format string   output format (json, csv, sarif)
  -r, --report-path string     report file

>   -b, --baseline-path          path to existing a previously generated report with known issues that gitleaks should ignore

Reviewer: remove "existing"?

**NOTE**: the `protect` command can only be used on git repos, running `protect` on files or directories will result in an error message.

### Creating a baseline

When scanning large repositories or repositories with a long history, it can be convenient to use a baseline. When using a baseline, 

> gitleaks will ignore any old findings which is present in the baseline. To create a baseline, run gitleaks with the `--report-path` parameter. 

Reviewer 2: ```suggestion
gitleaks will ignore any old findings that are present in the baseline. A baseline can be any gitleaks report. To create a gitleaks report, run gitleaks with the `--report-path` parameter. 
```

```

> gitleaks detect --report-path .gitleaksbaseline # This will save the report in a file called .gitleaksbaseline

Reviewer 2: I think we should rename the report file to something more conventional like `gitleaks-report.json`. Naming it `.gitleaksbaseline` might give readers the misconception that `--report-path` is something other than a gitleaks report.

```

> ***NOTE***: If you decide to save the baseline within the source three, you must add it to the .gitleaksignore file. Otherwise, the contents of this file will generate new issues.

Reviewer 2: ```suggestion
***NOTE***: If you decide to commit the baseline file to the repo, you must add it to the .gitleaksignore file. Otherwise, the contents of the baseline file will generate new issues.
```

Once as baseline is created it can be applied when running the detect command again:
```

> gitleaks detect --baseline-path .gitleaksbaseline --report-path findings.json

Reviewer 2: Changes should be reflected here too

**NOTE**: the `protect` command can only be used on git repos, running `protect` on files or directories will result in an error message.

### Creating a baseline

When scanning large repositories or repositories with a long history, it can be convenient to use a baseline. When using a baseline, 
gitleaks will ignore any old findings that are present in the baseline. A baseline can be any gitleaks report. To create a gitleaks report, run gitleaks with the `--report-path` parameter. 

```
gitleaks detect --report-path gitleaks-report.json # This will save the report in a file called gitleaks-report.json
```

> ***NOTE***: If you decide to commit the baseline file to the repo, you must add it to the .gitleaksignore file. Otherwise, the contents of the baseline file will generate new issues.

Reviewer 2: > ***NOTE***: If you decide to commit the baseline file to the repo, you must add it to the .gitleaksignore file. Otherwise, the contents of the baseline file will generate new issues.

@gawansch thanks for bearing with us on this review. I did some testing locally and it works great! However, I think this can be avoided by _automatically_ adding an `allowlist` entry that specifies the path to the `baseline`. We do this with `config`. See https://github.com/zricethezav/gitleaks/blob/master/detect/detect.go#L426-L429.

We could add the baseline path to that condition like:
```
diff --git a/detect/detect.go b/detect/detect.go
index ff1dcda..8494cf4 100644
--- a/detect/detect.go
+++ b/detect/detect.go
@@ -69,6 +69,9 @@ type Detector struct {
        // a list of known findings that should be ignored
        baseline []report.Finding

+       // path to baseline
+       baselinePath string
+
        // gitleaksIgnore
        gitleaksIgnore map[string]bool
 }
@@ -156,6 +159,7 @@ func (d *Detector) AddBaseline(baselinePath string) error {
                }
                d.baseline = baseline
        }
+       d.baselinePath = baselinePath
        return nil
 }

@@ -438,7 +442,7 @@ func (d *Detector) Detect(fragment Fragment) []report.Finding {

        // check if filepath is allowed
        if fragment.FilePath != "" && (d.Config.Allowlist.PathAllowed(fragment.FilePath) ||
-               fragment.FilePath == d.Config.Path) {
+               fragment.FilePath == d.Config.Path || fragment.FilePath == d.baselinePath) {
                return findings
        }

```

Author: Good point! It will work as long as the baseline path is present, and improve the usability. 

When generating a new baseline users will do that without the `baseline-path` argument. The new baseline will contain these extra entries, which can lead to some confusion, but at this point we can count on users being able to put two and two together.

Reviewer: > When generating a new baseline users will do that without the baseline-path argument. The new baseline will contain these extra entries, which can lead to some confusion

This is a good point. Once this feature gets a little mileage, hopefully we'll develop a best practice to avoid this. It might be as simple as "be sure to delete or move your old baseline before creating a new one".
